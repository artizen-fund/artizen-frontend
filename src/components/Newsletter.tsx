import { useState, useEffect } from 'react'
import styled from 'styled-components'
import MailchimpSubscribe, { FormHooks, NameFormFields } from 'react-mailchimp-subscribe'
import { rgba } from '@lib'
import Link from 'next/link'
import { Button, PagePadding, Icon } from '@components'
import { breakpoint, palette, typography } from '@theme'
import { newsletter } from '@copy/common'

const Newsletter = ({ subscribe }: FormHooks<NameFormFields>) => {
  return (
    <StyledPagePadding black>
      <Wrapper>
        <Copy>
          <Header>{newsletter.headline}</Header>
          <Subhead>{newsletter.subhead}</Subhead>
        </Copy>
        <Link href="https://artizenfund.substack.com/">
          <StyledButton onClick={() => {}} inverted level={0}>
            {newsletter.buttonLabel}
          </StyledButton>
        </Link>
      </Wrapper>
    </StyledPagePadding>
  )
}

const StyledPagePadding = styled(props => <PagePadding {...props} />)``

const StyledButton = styled(props => <Button {...props} />)`
  grid-area: submit;
  height: 24px;
`

const Wrapper = styled.div`
  display: grid;
  align-items: center;
  gap: 10px;
  grid-template-areas:
    'copy copy'
    'submit submit';

  @media only screen and (min-width: ${breakpoint.laptop}px) {
    gap: 12px;
    grid-template-areas: 'copy copy submit submit';
  }
  @media only screen and (min-width: ${breakpoint.desktop}px) {
    gap: 16px;
  }
`

const Copy = styled.div`
  grid-area: copy;
`

const Header = styled.div`
  ${typography.title.l2}
  color: ${rgba(palette.white)};
`

const Subhead = styled.div`
  margin-top: 0.65em;
  ${typography.label.l1}
  color: ${rgba(palette.barracuda)};
`

const MailChimpForm = () => {
  return <MailchimpSubscribe render={props => <Newsletter {...props} />} />
}

export default MailChimpForm
