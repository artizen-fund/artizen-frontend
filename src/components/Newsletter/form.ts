import { JsonSchema } from '@jsonforms/core'

export const schema: JsonSchema = {
  type: 'object',
  properties: {
    FNAME: {
      type: 'string',
      minLength: 2,
    },
    LNAME: {
      type: 'string',
      minLength: 2,
    },
    EMAIL: {
      type: 'string',
      format: 'email',
    },
    OPTIN: { type: 'boolean' },
  },
  required: ['EMAIL', 'OPTIN'],
}

export interface FormState extends Record<string, unknown> {
  FNAME?: string
  LNAME?: string
  EMAIL?: string
  OPTIN?: boolean
}

export const initialState: FormState = { OPTIN: true }

export const uischema = {
  type: 'VerticalLayout',
  elements: [
    {
      type: 'Control',
      scope: '#/properties/FNAME',
      label: 'First Name',
    },
    {
      type: 'Control',
      scope: '#/properties/LNAME',
      label: 'Last Name',
    },
    {
      type: 'Control',
      scope: '#/properties/EMAIL',
      label: 'Email Address',
    },
    {
      type: 'Control',
      scope: '#/properties/OPTIN',
      label: 'Opt-in to receive the Artizen newsletter.',
      options: { inverted: true },
    },
  ],
}
