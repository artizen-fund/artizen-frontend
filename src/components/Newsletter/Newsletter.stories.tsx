import Newsletter from './'

export default {
  title: 'components/Newsletter',
  component: Newsletter,
}

export const NewsletterComponent = () => {
  return <Newsletter />
}
