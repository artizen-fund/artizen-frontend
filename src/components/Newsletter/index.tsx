import { useState } from 'react'
import styled from 'styled-components'
import { breakpoint, palette } from '@theme'
import { rgba } from '@lib'
import { Form, Button } from '@components'
import { schema, uischema, initialState } from './form'

const Newsletter = () => {
  const [submitted, setSubmitted] = useState(false)
  return (
    <Wrapper className={submitted ? 'submitted' : ''}>
      <Instructions>
        <p>Join us in building the world's largest web3 fund for public goods</p>
        <p>Sign up for our free newsletter</p>
      </Instructions>
      <Form {...{ schema, uischema, initialState }} />
      <StyledButton onClick={() => setSubmitted(!submitted)}>Submit</StyledButton>
      <Thanks>Thank you Thank you</Thanks>
    </Wrapper>
  )
}

const Thanks = styled.div`
  grid-area: g;
`

const Wrapper = styled.div`
  display: grid;
  gap: 20px;
  grid-template-areas: 'a a' 'b b' 'c d' 'e e' 'f f';
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    grid-template-areas: 'a a c d' 'a a e e' 'b b f f';
  }
  &.submitted {
    grid-template-areas: 'a a' 'b b' 'g g' 'g g' 'g g';
    @media only screen and (min-width: ${breakpoint.laptop}px) {
      grid-template-areas: 'a a g g' 'a a g g' 'b b g g';
    }
  }

  .vertical-layout,
  .vertical-layout-item {
    display: contents;
  }

  *[id='#/properties/email'] {
    grid-area: e;
  }

  *[id='#/properties/firstName'] {
    grid-area: c;
  }

  *[id='#/properties/lastName'] {
    grid-area: d;
  }

  *[id='#/properties/optIn'] {
    grid-area: b;
  }

  &.submitted {
    *[id='#/properties/email'],
    *[id='#/properties/firstName'],
    *[id='#/properties/lastName'],
    *[id='#/properties/optIn'] {
      display: none;
    }
  }
`

const Instructions = styled.div`
  grid-area: a;
`

const StyledButton = styled(props => <Button {...props} />)`
  grid-area: f;
`

export default Newsletter
