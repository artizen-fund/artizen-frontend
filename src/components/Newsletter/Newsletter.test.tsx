import '@testing-library/jest-dom/extend-expect'
import { render } from '@testing-library/react'
import Newsletter from './'

describe('Newsletter', () => {
  it('renders Newsletter unchanged', () => {
    const { container } = render(<Newsletter />)
    expect(container).toMatchSnapshot()
  })
})
