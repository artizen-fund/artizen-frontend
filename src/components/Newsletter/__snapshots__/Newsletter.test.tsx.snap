// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newsletter renders Newsletter unchanged 1`] = `
<div>
  <div
    class="sc-85485da3-0 eEcuNe"
  >
    <p>
      Join us in building the world's largest web3 fund for public goods
    </p>
    <p>
      Sign up for our free newsletter
    </p>
    <div
      class="sc-3d351cf1-0 cuONLG"
    >
      <div
        class="vertical-layout"
      >
        <div
          class="vertical-layout-item"
        >
          <div
            class="horizontal-layout"
          >
            <div
              class="horizontal-layout-item horizontal-layout-2"
            >
              <div
                class="sc-e191228e-0 bwBNFU"
              >
                <div
                  class="sc-9db10430-0 gtsrND"
                >
                  <input
                    class="noData"
                    minlength="2"
                    type="text"
                    value=""
                  />
                  <label
                    class="sc-aec1a1a-0 fjKCPb"
                  >
                    First Name
                  </label>
                </div>
                <div
                  class="sc-955ed11e-0 eSGmlb"
                />
              </div>
            </div>
            <div
              class="horizontal-layout-item horizontal-layout-2"
            >
              <div
                class="sc-e191228e-0 bwBNFU"
              >
                <div
                  class="sc-9db10430-0 gtsrND"
                >
                  <input
                    class="noData"
                    minlength="2"
                    type="text"
                    value=""
                  />
                  <label
                    class="sc-aec1a1a-0 fjKCPb"
                  >
                    Last Name
                  </label>
                </div>
                <div
                  class="sc-955ed11e-0 eSGmlb"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="vertical-layout-item"
        >
          <div
            class="sc-e191228e-0 gezOsv"
          >
            <div
              class="sc-9db10430-0 gtsrND"
            >
              <input
                class="noData"
                required=""
                type="text"
                value=""
              />
              <label
                class="sc-aec1a1a-0 fjKCPb"
              >
                Email Address
                 *
              </label>
            </div>
            <div
              class="sc-955ed11e-0 eSGmlb hasErrors"
            >
              is a required property
            </div>
          </div>
        </div>
        <div
          class="vertical-layout-item"
        >
          <label
            class="sc-94a2615b-4 vyhjA"
          >
            <div
              class="sc-94a2615b-0 dEeEHB"
            >
              <input
                class="sc-94a2615b-1 foJfOn"
                required=""
                type="checkbox"
              />
              <span
                class="sc-94a2615b-2 gZzYJ"
              />
            </div>
            <span
              class="sc-94a2615b-3 iLipNt"
            >
              Opt-in to receive the Artizen newsletter.
            </span>
          </label>
        </div>
      </div>
      <button
        class="sc-a1052e44-2 frqjg"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;
